test/server/http/admin_test.cc

```diff
@@ -75,19 +81,45 @@ class AdminInstanceTest : public testing::TestWithParam<Network::Address::IpVers
   AdminInstanceTest()
       : address_out_path_(TestEnvironment::temporaryPath("admin.address")),
         cpu_profile_path_(TestEnvironment::temporaryPath("envoy.prof")),
+        log_level_setter_(spdlog::level::warn),
         admin_("/dev/null", cpu_profile_path_, address_out_path_,
                Network::Test::getCanonicalLoopbackAddress(GetParam()), server_,
-               listener_scope_.createScope("listener.admin.")) {
+               listener_scope_.createScope("listener.admin.")),
+        expect_no_logs_(true) {
+
     EXPECT_EQ(std::chrono::milliseconds(100), admin_.drainTimeout());
     admin_.tracingStats().random_sampling_.inc();
     EXPECT_TRUE(admin_.setCurrentClientCertDetails().empty());
   }
 
+  void TearDown() override {
+    if (expect_no_logs_) {
+      EXPECT_EQ(0, mock_logger_.messages().size());
+    }
+  }
+
+  std::string recordedLog() {
+    const std::vector<std::string> logs = mock_logger_.messages();
+    if (!logs.empty()) {
+      return logs[0];
+    }
+    return "";
+  }
+
+  bool recordedLogContains(absl::string_view substr) {
```

Comment:
Would it be possible to get a model similar to `EXPECT_CALL`, e.g. `EXPECT_LOG('Some log message")` with this new infra?

Reply:
Excellent idea -- I think this simplifies a lot of things.  Done.  @mattklein123 does this address your concerns about having the command-line flags mostly work, except in the context of EXPECT_LOG_CONTAINS?

Reply:
Sure it's definitely better, though I honestly still would prefer some type of forking logger that can be used during tests. 

If a spdlog logger has multiple sinks, can the level be set independently on each sink? Just wondering if for test code we could install a forking logger, and then independently set the forked logger to warn and greater only, which realistically I think is all we are ever going to test?

If we think this is the best we can do I can live with it, but it still seems mildly hacky IMHO.

Reply:
spdlog::logger contains the level, not the sinks, unfortunately.

Reply:
But in a sink we can get the level, right? https://github.com/gabime/spdlog/blob/master/include/spdlog/details/log_msg.h#L16

I guess I'm just thinking of a solution that looks like:
1) Forking logger in tests.
2) If a test wants to test logging, it says the log level it wants to test at.
3) We set actual log level at logger to min<CLI, requested>
4) Enforce log level in the test sinks?